File: RainLanguage/RainLanguage/Compiler/FileSpaceLink.cpp
#include "FileSpace.h"
#include "Context.h"
#include "../KernelDeclarations.h"

#define FIND_DECLARATION(type) declarations.Clear();context.FindDeclaration(manager, type.name, declarations); Type findType = type.GetType(manager, declarations);

Type FileType::GetType(DeclarationManager* manager, const List<CompilingDeclaration, true>& declarations)
{
	if (declarations.Count() == 0)MESSAGE2(manager->messages, name, MessageType::ERROR_DECLARATION_NOT_FOUND)
	else if (declarations.Count() == 1)
	{
		const CompilingDeclaration& declaration = declarations[0];
		if (declaration.category == DeclarationCategory::Enum) return Type(declaration.library, TypeCode::Enum, declaration.index, dimension);
		else if (declaration.category == DeclarationCategory::Struct)return Type(declaration.library, TypeCode::Struct, declaration.index, dimension);
		else if (declaration.category == DeclarationCategory::Class)return Type(declaration.library, TypeCode::Handle, declaration.index, dimension);
		else if (declaration.category == DeclarationCategory::Interface)return Type(declaration.library, TypeCode::Interface, declaration.index, dimension);
		else if (declaration.category == DeclarationCategory::Delegate)return Type(declaration.library, TypeCode::Delegate, declaration.index, dimension);
		else if (declaration.category == DeclarationCategory::Task)return Type(declaration.library, TypeCode::Task, declaration.index, dimension);
		else
		{
			MESSAGE3(manager->messages, name, MessageType::ERROR_NOT_TYPE_DECLARATION, manager->GetDeclaration(declaration)->GetFullName(manager->stringAgency));
			return Type();
		}
	}
	else for (uint32 i = 0; i < declarations.Count(); i++)
		MESSAGE3(manager->messages, name, MessageType::ERROR_DECLARATION_EQUIVOCAL, manager->GetDeclaration(declarations[i])->GetFullName(manager->stringAgency));
	return Type();
}

bool TryGetDeclaration(FileSpace* space, const Anchor& name, CompilingDeclaration& declaration)
{
	List<CompilingDeclaration, true>* declarations = NULL;
	if (space->compiling->abstract->declarations.TryGet(name.content, declarations) && declarations->Count())
	{
		declaration = (*declarations)[0];
		return true;
	}
	return false;
}

void FileSpace::Link(DeclarationManager* manager, List<List<AbstractSpace*, true>*, true>* relySpaceCollector)
{
	for (uint32 i = 0; i < children.Count(); i++) children[i]->Link(manager, relySpaceCollector);
	List<AbstractSpace*, true>* relies = new List<AbstractSpace*, true>(relyCompilingSpaces.Count() + relySpaces.Count());
	relySpaceCollector->Add(relies);
	relies->Add(relySpaces);
	for (uint32 i = 0; i < relyCompilingSpaces.Count(); i++)
		relies->Add(relyCompilingSpaces[i]->abstract);
	Context context = Context(compiling, relies);
	List<CompilingDeclaration, true> declarations = List<CompilingDeclaration, true>(0);
	for (uint32 i = 0; i < variables.Count(); i++)
	{
		FileVariable* file = &variables[i];
		CompilingVariable* compilingVariable = manager->compilingLibrary.variables[file->index];
		AbstractVariable* abstractVariable = manager->selfLibaray->variables[file->index];
		compilingVariable->relies = relies;
		FIND_DECLARATION(file->type);
		abstractVariable->type = compilingVariable->type = findType;
	}
	for (uint32 x = 0; x < functions.Count(); x++)
	{
		FileFunction* file = &functions[x];
		CompilingFunction* compilingFunction = manager->compilingLibrary.functions[file->index];
		AbstractFunction* abstractFunction = manager->selfLibaray->functions[file->index];
		for (uint32 y = 0; y < file->parameters.Count(); y++)
		{
			FileParameter* parameter = &file->parameters[y];
			FIND_DECLARATION(parameter->type);
			new (compilingFunction->parameters.Add())CompilingFunctionDeclaration::Parameter(parameter->name, findType);
			abstractFunction->parameters.AddElement(findType, 0);
		}
		for (uint32 y = 0; y < file->returns.Count(); y++)
		{
			FIND_DECLARATION(file->returns[y]);
			compilingFunction->returns.Add(findType);
			abstractFunction->returns.AddElement(findType, 0);
		}
		compilingFunction->relies = relies;
	}
	for (uint32 x = 0; x < enums.Count(); x++) manager->compilingLibrary.enums[enums[x].index]->relies = relies;
	for (uint32 x = 0; x < structs.Count(); x++)
	{
		FileStruct* file = &structs[x];
		CompilingStruct* compilingStruct = manager->compilingLibrary.structs[file->index];
		AbstractStruct* abstractStruct = manager->selfLibaray->structs[file->index];
		CompilingDeclaration fileDeclaratioin;
		if (!TryGetDeclaration(file->space, file->name, fileDeclaratioin)) EXCEPTION("自己的命名空间中找不到自己的定义");
		ASSERT_DEBUG(fileDeclaratioin.category == DeclarationCategory::Struct, "定义类型错误");
		Type fileType = Type(LIBRARY_SELF, TypeCode::Struct, fileDeclaratioin.index, 0);
		for (uint32 y = 0; y < file->variables.Count(); y++)
		{
			FileStruct::Variable* member = &file->variables[y];
			FIND_DECLARATION(member->type);
			abstractStruct->variables[y]->type = compilingStruct->variables[y]->type = findType;
		}
		for (uint32 y = 0; y < file->functions.Count(); y++)
		{
			FileFunction* member = &file->functions[y];
			CompilingFunction* compilingMember = manager->compilingLibrary.functions[compilingStruct->functions[y]];
			AbstractFunction* abstractMember = manager->selfLibaray->functions[compilingStruct->functions[y]];

			new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(file->name, fileType);
			abstractMember->parameters.AddElement(fileType, 0);
			for (uint32 z = 0; z < member->parameters.Count(); z++)
			{
				FileParameter* parameter = &member->parameters[z];
				FIND_DECLARATION(parameter->type);
				new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(parameter->name, findType);
				abstractMember->parameters.AddElement(findType, 0);
			}
			for (uint32 z = 0; z < member->returns.Count(); z++)
			{
				FIND_DECLARATION(member->returns[z]);
				compilingMember->returns.Add(findType);
				abstractMember->returns.AddElement(findType, 0);
			}
			compilingMember->relies = relies;
		}
	}
	for (uint32 x = 0; x < classes.Count(); x++)
	{
		FileClass* file = &classes[x];
		CompilingClass* compilingClass = manager->compilingLibrary.classes[file->index];
		AbstractClass* abstractClass = manager->selfLibaray->classes[file->index];
		CompilingDeclaration fileDeclaratioin;
		if (!TryGetDeclaration(file->space, file->name, fileDeclaratioin)) EXCEPTION("自己的命名空间中找不到自己的定义");
		ASSERT_DEBUG(fileDeclaratioin.category == DeclarationCategory::Class, "定义类型错误");
		Type fileType = Type(LIBRARY_SELF, TypeCode::Handle, fileDeclaratioin.index, 0);
		if (file->parent.Count())
		{
			FIND_DECLARATION(FileType(file->parent, 0));
			if (findType.code == TypeCode::Interface)
			{
				abstractClass->parent = compilingClass->parent = TYPE_Handle;
				compilingClass->inherits.Add(findType);
				abstractClass->inherits.Add(findType);
			}
			else abstractClass->parent = compilingClass->parent = findType;
		}
		else abstractClass->parent = compilingClass->parent = TYPE_Handle;
		for (uint32 y = 0; y < file->inherits.Count(); y++)
		{
			FIND_DECLARATION(file->inherits[y]);
			if (findType.library != INVALID)
			{
				uint32 index = compilingClass->inherits.IndexOf(findType);
				if (index == INVALID)
				{
					compilingClass->inherits.Add(findType);
					abstractClass->inherits.Add(findType);
				}
				else MESSAGE2(manager->messages, file->inherits[y].name, MessageType::ERROR_DUPLICATE_INHERITANCE);
			}
		}
		if (file->constructors.Count())
		{
			for (uint32 y = 0; y < file->constructors.Count(); y++)
			{
				FileClass::Constructor* member = &file->constructors[y];
				CompilingFunction* compilingMember = manager->compilingLibrary.functions[compilingClass->constructors[y]->function];
				AbstractFunction* abstractMember = manager->selfLibaray->functions[compilingClass->constructors[y]->function];
				new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(file->name, fileType);
				abstractMember->parameters.AddElement(fileType, 0);
				for (uint32 z = 0; z < member->parameters.Count(); z++)
				{
					FileParameter* parameter = &member->parameters[z];
					FIND_DECLARATION(parameter->type);
					new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(parameter->name, findType);
					abstractMember->parameters.AddElement(findType, 0);
				}
				compilingMember->returns.Add(fileType);
				abstractMember->returns.AddElement(fileType, 0);
				compilingMember->relies = relies;
			}
		}
		else
		{
			CompilingFunction* compilingMember = manager->compilingLibrary.functions[compilingClass->constructors[0]->function];
			AbstractFunction* abstractMember = manager->selfLibaray->functions[compilingClass->constructors[0]->function];
			new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(file->name, fileType);
			abstractMember->parameters.AddElement(fileType, 0);
			compilingMember->returns.Add(fileType);
			abstractMember->returns.AddElement(fileType, 0);
			compilingMember->relies = relies;
		}
		for (uint32 y = 0; y < file->variables.Count(); y++)
		{
			FileClass::Variable* member = &file->variables[y];
			FIND_DECLARATION(member->type);
			abstractClass->variables[y]->type = compilingClass->variables[y]->type = findType;
		}
		for (uint32 y = 0; y < file->functions.Count(); y++)
		{
			FileFunction* member = &file->functions[y];
			CompilingFunction* compilingMember = manager->compilingLibrary.functions[compilingClass->functions[y]];
			AbstractFunction* abstractMember = manager->selfLibaray->functions[compilingClass->functions[y]];
			new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(file->name, fileType);
			abstractMember->parameters.AddElement(fileType, 0);
			for (uint32 z = 0; z < member->parameters.Count(); z++)
			{
				FileParameter* parameter = &member->parameters[z];
				FIND_DECLARATION(parameter->type);
				new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(parameter->name, findType);
				abstractMember->parameters.AddElement(findType, 0);
			}
			for (uint32 z = 0; z < member->returns.Count(); z++)
			{
				FIND_DECLARATION(member->returns[z]);
				compilingMember->returns.Add(findType);
				abstractMember->returns.AddElement(findType, 0);
			}
			compilingMember->relies = relies;
		}
		compilingClass->relies = relies;
	}
	for (uint32 x = 0; x < interfaces.Count(); x++)
	{
		FileInterface* file = &interfaces[x];
		CompilingDeclaration fileDeclaratioin;
		if (!TryGetDeclaration(file->space, file->name, fileDeclaratioin)) EXCEPTION("自己的命名空间中找不到自己的定义");
		ASSERT_DEBUG(fileDeclaratioin.category == DeclarationCategory::Interface, "定义类型错误");
		Type fileType = Type(LIBRARY_SELF, TypeCode::Interface, fileDeclaratioin.index, 0);
		CompilingInterface* compilingInterface = manager->compilingLibrary.interfaces[file->index];
		AbstractInterface* abstractInterface = manager->selfLibaray->interfaces[file->index];
		for (uint32 y = 0; y < file->inherits.Count(); y++)
		{
			FIND_DECLARATION(file->inherits[y]);
			if (findType.library != INVALID)
			{
				uint32 index = compilingInterface->inherits.IndexOf(findType);
				if (index == INVALID)
				{
					compilingInterface->inherits.Add(findType);
					abstractInterface->inherits.Add(findType);
				}
				else MESSAGE2(manager->messages, file->inherits[y].name, MessageType::ERROR_DUPLICATE_INHERITANCE);
			}
		}
		for (uint32 y = 0; y < file->functions.Count(); y++)
		{
			FileInterface::Function* member = &file->functions[y];
			CompilingInterface::Function* compilingMember = compilingInterface->functions[y];
			AbstractFunction* abstractMember = abstractInterface->functions[y];
			new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(file->name, fileType);
			abstractMember->parameters.AddElement(fileType, 0);
			for (uint32 z = 0; z < member->parameters.Count(); z++)
			{
				FIND_DECLARATION(member->parameters[z].type);
				new (compilingMember->parameters.Add())CompilingFunctionDeclaration::Parameter(member->parameters[z].name, findType);
				abstractMember->parameters.AddElement(findType, 0);
			}
			for (uint32 z = 0; z < member->returns.Count(); z++)
			{
				FIND_DECLARATION(member->returns[z]);
				compilingMember->returns.Add(findType);
				abstractMember->returns.AddElement(findType, 0);
			}
		}
	}
	for (uint32 x = 0; x < delegates.Count(); x++)
	{
		FileDelegate* file = &delegates[x];
		CompilingDelegate* compilingDelegate = manager->compilingLibrary.delegates[file->index];
		AbstractDelegate* abstractDelegate = manager->selfLibaray->delegates[file->index];
		for (uint32 y = 0; y < file->parameters.Count(); y++)
		{
			FIND_DECLARATION(file->parameters[y].type);
			new (compilingDelegate->parameters.Add())CompilingFunctionDeclaration::Parameter(file->parameters[y].name, findType);
			abstractDelegate->parameters.AddElement(findType, 0);
		}
		for (uint32 y = 0; y < file->returns.Count(); y++)
		{
			FIND_DECLARATION(file->returns[y]);
			compilingDelegate->returns.Add(findType);
			abstractDelegate->returns.AddElement(findType, 0);
		}
	}
	for (uint32 x = 0; x < tasks.Count(); x++)
	{
		FileTask* file = &tasks[x];
		CompilingTask* compilingTask = manager->compilingLibrary.tasks[file->index];
		AbstractTask* abstractTask = manager->selfLibaray->tasks[file->index];
		for (uint32 y = 0; y < file->returns.Count(); y++)
		{
			FIND_DECLARATION(file->returns[y]);
			compilingTask->returns.Add(findType);
			abstractTask->returns.AddElement(findType, 0);
		}
	}
	for (uint32 x = 0; x < natives.Count(); x++)
	{
		FileNative* file = &natives[x];
		CompilingNative* compilingNative = manager->compilingLibrary.natives[file->index];
		AbstractNative* abstractNative = manager->selfLibaray->natives[file->index];
		for (uint32 y = 0; y < file->parameters.Count(); y++)
		{
			FIND_DECLARATION(file->parameters[y].type);
			new (compilingNative->parameters.Add())CompilingFunctionDeclaration::Parameter(file->parameters[y].name, findType);
			abstractNative->parameters.AddElement(findType, 0);
		}
		for (uint32 y = 0; y < file->returns.Count(); y++)
		{
			FIND_DECLARATION(file->returns[y]);
			compilingNative->returns.Add(findType);
			abstractNative->returns.AddElement(findType, 0);
		}
	}
}

FileSpace::~FileSpace()
{
	for (uint32 i = 0; i < children.Count(); i++) delete children[i];
	children.Clear();
}

File: RainLanguageServer/RainLanguage/FileSpaceLink.cs
using RainLanguageServer.RainLanguage.GrammaticalAnalysis;

namespace RainLanguageServer.RainLanguage
{
    internal partial class FileSpace
    {
        private Type GetType(Context context, ASTManager manager, FileType fileType)
        {
            Type result = default;
            if (context.TryFindDeclaration(manager, fileType.name, out var declarations, collector))
            {
                if (declarations.Count == 1)
                {
                    result = declarations[0].declaration.GetDefineType();
                    if (!result.Vaild) collector.Add(fileType.name, CErrorLevel.Error, "无效的类型");
                }
                else
                {
                    var msg = new CompileMessage(fileType.name, CErrorLevel.Error, "类型不明确");
                    foreach (var item in declarations)
                        msg.related.Add(new RelatedInfo(item.name, item.GetFullName()));
                    collector.Add(msg);
                }
            }
            else collector.Add(fileType.name, CErrorLevel.Error, "声明未找到");
            if (manager.GetSourceDeclaration(result) is CompilingDeclaration target)
                target.references.Add(fileType.GetNameRange());
            return result.GetDimensionType(fileType.dimension);
        }
        public void Link(ASTManager manager, CompilingLibrary library, bool cite)
        {
            foreach (var child in children) child.Link(manager, library, cite);
            var context = new Context(document, compiling, relies, default);
            foreach (var file in variables)
            {
                var type = GetType(context, manager, file.type);
                var declaration = new Declaration(library.name, file.visibility, DeclarationCategory.Variable, compiling.GetChildName(file.name.ToString()), default);
                var variable = new CompilingVariable(file.name, declaration, file.attributes, compiling, cite ? file : null, file.isReadonly, type, file.expression, relies);
                if (file.isReadonly && type != Type.BOOL && type != Type.BYTE && type != Type.CHAR && type != Type.INT && type != Type.REAL && type != Type.STRING)
                    file.space.collector.Add(file.name, CErrorLevel.Error, "该类型不能作为常量");
                library.variables.Add(variable);
                file.compiling = variable;
                compiling.AddDeclaration(variable);
            }
            foreach (var file in functions)
            {
                var parameters = new List<CompilingCallable.Parameter>();
                foreach (var parameter in file.parameters)
                    parameters.Add(new CompilingCallable.Parameter(parameter.name, GetType(context, manager, parameter.type)));
                var parameterTypes = new List<Type>();
                foreach (var parameter in parameters) parameterTypes.Add(parameter.type);
                var returnTypes = new List<Type>();
                foreach (var type in file.returns)
                    returnTypes.Add(GetType(context, manager, type));
                var declaration = new Declaration(library.name, file.visibility, DeclarationCategory.Function, compiling.GetChildName(file.name.ToString()), new Tuple(parameterTypes));
                var function = new CompilingFunction(file.name, declaration, file.attributes, compiling, cite ? file : null, parameters, new Tuple(returnTypes), new LogicBlock(document, null, parameters, returnTypes, compiling, file.body, relies, collector));
                library.functions.Add(function);
                file.compiling = function;
                compiling.AddDeclaration(function);
            }
            foreach (var file in enums)
            {
                compiling.declarations.TryGetValue(file.name.ToString(), out var declarations);
                var compilingEnum = (CompilingEnum)declarations!.Find(value => value.name == file.name)!;

                foreach (var element in file.elements)
                {
                    var declaration = new Declaration(library.name, Visibility.Public, DeclarationCategory.EnumElement, compiling.GetMemberName(file.name.ToString(), element.name.ToString()), default);
                    var compilingEnumElement = new CompilingEnum.Element(element.name, declaration, cite ? element : null);
                    compilingEnum.elements.Add(compilingEnumElement);
                }
            }
            foreach (var file in structs)
            {
                compiling.declarations.TryGetValue(file.name.ToString(), out var declarations);
                var compilingStruct = (CompilingStruct)declarations!.Find(value => value.name == file.name)!;

                foreach (var variable in file.variables)
                {
                    var type = GetType(context, manager, variable.type);
                    var declaration = new Declaration(library.name, variable.visibility, DeclarationCategory.StructVariable, compiling.GetMemberName(file.name.ToString(), variable.name.ToString()), default);
                    var compilingVariable = new CompilingVariable(variable.name, declaration, variable.attributes, compiling, cite ? variable : null, variable.isReadonly, type, variable.expression, relies);
                    compilingStruct.variables.Add(compilingVariable);
                    variable.compiling = compilingVariable;
                }
                foreach (var function in file.functions)
                {
                    var parameters = new List<CompilingCallable.Parameter>();
                    foreach (var parameter in function.parameters)
                        parameters.Add(new CompilingCallable.Parameter(parameter.name, GetType(context, manager, parameter.type)));
                    var parameterTypes = new List<Type>();
                    foreach (var parameter in parameters) parameterTypes.Add(parameter.type);
                    var returnTypes = new List<Type>();
                    foreach (var type in function.returns)
                        returnTypes.Add(GetType(context, manager, type));
                    var declaration = new Declaration(library.name, function.visibility, DeclarationCategory.StructFunction, compiling.GetMemberName(file.name.ToString(), function.name.ToString()), new Tuple(parameterTypes));
                    var compilingFunction = new CompilingFunction(function.name, declaration, function.attributes, compiling, cite ? function : null, parameters, new Tuple(returnTypes), new LogicBlock(document, compilingStruct, parameters, returnTypes, compiling, function.body, relies, collector));
                    compilingStruct.functions.Add(compilingFunction);
                    function.compiling = compilingFunction;
                }
            }
            foreach (var file in interfaces)
            {
                compiling.declarations.TryGetValue(file.name.ToString(), out var declarations);
                var compilingInterface = (CompilingInterface)declarations!.Find(value => value.name == file.name)!;

                foreach (var fileType in file.inherits)
                {
                    var type = GetType(context, manager, fileType);
                    if (type.dimension > 0) collector.Add(fileType.name, CErrorLevel.Error, "不能继承数组");
                    else if (type.code != TypeCode.Interface && type != Type.HANDLE) collector.Add(fileType.name, CErrorLevel.Error, "必须是接口");
                    compilingInterface.inherits.Add(type);
                    if (manager.GetSourceDeclaration(type) is CompilingInterface inherit)
                        compilingInterface.implements.Add(inherit);
                }
                foreach (var function in file.functions)
                {
                    var parameters = new List<CompilingCallable.Parameter>();
                    foreach (var parameter in function.parameters)
                        parameters.Add(new CompilingCallable.Parameter(parameter.name, GetType(context, manager, parameter.type)));
                    var parameterTypes = new List<Type>();
                    foreach (var parameter in parameters) parameterTypes.Add(parameter.type);
                    var returnTypes = new List<Type>();
                    foreach (var type in function.returns)
                        returnTypes.Add(GetType(context, manager, type));
                    var declaration = new Declaration(library.name, function.visibility, DeclarationCategory.InterfaceFunction, compiling.GetMemberName(file.name.ToString(), function.name.ToString()), new Tuple(parameterTypes));
                    var compilingFunction = new CompilingAbstractFunction(function.name, declaration, function.attributes, compiling, cite ? function : null, parameters, new Tuple(returnTypes));
                    compilingInterface.callables.Add(compilingFunction);
                    function.compiling = compilingFunction;
                }
            }
            foreach (var file in classes)
            {
                compiling.declarations.TryGetValue(file.name.ToString(), out var declarations);
                var compilingClass = (CompilingClass)declarations!.Find(value => value.name == file.name)!;

                for (int i = 0; i < file.inherits.Count; i++)
                {
                    var type = GetType(context, manager, file.inherits[i]);
                    if (type.dimension > 0) collector.Add(file.inherits[i].name, CErrorLevel.Error, "不能继承数组");
                    if (i > 0 || type.code == TypeCode.Interface)
                    {
                        compilingClass.inherits.Add(type);
                        if (type.code != TypeCode.Interface) collector.Add(file.inherits[i].name, CErrorLevel.Error, "必须是接口");
                    }
                    else
                    {
                        compilingClass.parent = type;
                        if (type.code != TypeCode.Handle) collector.Add(file.inherits[i].name, CErrorLevel.Error, "不能继承该类型");
                    }
                    var inherit = manager.GetSourceDeclaration(type);
                    if (inherit is CompilingInterface @interface) @interface.implements.Add(compilingClass);
                    else if (inherit is CompilingClass @class) @class.implements.Add(compilingClass);
                }
                foreach (var variable in file.variables)
                {
                    var type = GetType(context, manager, variable.type);
                    var declaration = new Declaration(library.name, variable.visibility, DeclarationCategory.ClassVariable, compiling.GetMemberName(file.name.ToString(), variable.name.ToString()), default);
                    var compilingVariable = new CompilingVariable(variable.name, declaration, variable.attributes, compiling, cite ? variable : null, variable.isReadonly, type, variable.expression, relies);
                    compilingClass.variables.Add(compilingVariable);
                    variable.compiling = compilingVariable;
                }
                foreach (var function in file.constructors)
                {
                    var parameters = new List<CompilingCallable.Parameter>();
                    foreach (var parameter in function.parameters)
                        parameters.Add(new CompilingCallable.Parameter(parameter.name, GetType(context, manager, parameter.type)));
                    var parameterTypes = new List<Type>();
                    foreach (var parameter in parameters) parameterTypes.Add(parameter.type);
                    var declaration = new Declaration(library.name, function.visibility, DeclarationCategory.Constructor, compiling.GetMemberName(file.name.ToString(), function.name.ToString()), new Tuple(parameterTypes));
                    var compilingFunction = new CompilingFunction(function.name, declaration, function.attributes, compiling, cite ? function : null, parameters, new Tuple([]), new LogicBlock(document, compilingClass, parameters, [], compiling, function.body, relies, collector));
                    compilingClass.constructors.Add(compilingFunction);
                    function.compiling = compilingFunction;
                }
                foreach (var function in file.functions)
                {
                    var parameters = new List<CompilingCallable.Parameter>();
                    foreach (var parameter in function.parameters)
                        parameters.Add(new CompilingCallable.Parameter(parameter.name, GetType(context, manager, parameter.type)));
                    var parameterTypes = new List<Type>();
                    foreach (var parameter in parameters) parameterTypes.Add(parameter.type);
                    var returnTypes = new List<Type>();
                    foreach (var type in function.returns)
                        returnTypes.Add(GetType(context, manager, type));
                    var declaration = new Declaration(library.name, function.visibility, DeclarationCategory.ClassFunction, compiling.GetMemberName(file.name.ToString(), function.name.ToString()), new Tuple(parameterTypes));
                    var compilingFunction = new CompilingVirtualFunction(function.name, declaration, function.attributes, compiling, cite ? function : null, parameters, new Tuple(returnTypes), new LogicBlock(document, compilingClass, parameters, returnTypes, compiling, function.body, relies, collector));
                    compilingClass.functions.Add(compilingFunction);
                    function.compiling = compilingFunction;
                }
            }
            foreach (var file in delegates)
            {
                compiling.declarations.TryGetValue(file.name.ToString(), out var declarations);
                var compilingDelegate = (CompilingDelegate)declarations!.Find(value => value.name == file.name)!;

                var name = compilingDelegate.declaration.name;
                var parameters = new List<CompilingCallable.Parameter>();
                foreach (var parameter in file.parameters)
                    parameters.Add(new CompilingCallable.Parameter(parameter.name, GetType(context, manager, parameter.type)));
                var parameterTypes = new List<Type>();
                foreach (var parameter in parameters) parameterTypes.Add(parameter.type);
                var returnTypes = new List<Type>();
                foreach (var type in file.returns)
                    returnTypes.Add(GetType(context, manager, type));
                var declaration = new Declaration(library.name, file.visibility, DeclarationCategory.Delegate, name, new Tuple(parameterTypes));

                declarations.Remove(compilingDelegate);
                library.delegates.Remove(compilingDelegate);

                var references = compilingDelegate.references;
                compilingDelegate = new CompilingDelegate(file.name, declaration, file.attributes, compiling, cite ? file : null, parameters, new Tuple(returnTypes));
                compilingDelegate.references.AddRange(references);

                declarations.Add(compilingDelegate);
                library.delegates.Add(compilingDelegate);
                file.compiling = compilingDelegate;
            }
            foreach (var file in tasks)
            {
                compiling.declarations.TryGetValue(file.name.ToString(), out var declarations);
                var compilingTask = (CompilingTask)declarations!.Find(value => value.name == file.name)!;

                var name = compilingTask.declaration.name;
                var returnTypes = new List<Type>();
                foreach (var type in file.returns)
                    returnTypes.Add(GetType(context, manager, type));
                var declaration = new Declaration(library.name, file.visibility, DeclarationCategory.Task, name, default);

                declarations.Remove(compilingTask);
                library.tasks.Remove(compilingTask);

                var references = compilingTask.references;
                compilingTask = new CompilingTask(file.name, declaration, file.attributes, compiling, cite ? file : null, new Tuple(returnTypes));
                compilingTask.references.AddRange(references);

                declarations.Add(compilingTask);
                library.tasks.Add(compilingTask);
                file.compiling = compilingTask;
            }
            foreach (var file in natives)
            {
                var parameters = new List<CompilingCallable.Parameter>();
                foreach (var parameter in file.parameters)
                    parameters.Add(new CompilingCallable.Parameter(parameter.name, GetType(context, manager, parameter.type)));
                var parameterTypes = new List<Type>();
                foreach (var parameter in parameters) parameterTypes.Add(parameter.type);
                var returnTypes = new List<Type>();
                foreach (var type in file.returns)
                    returnTypes.Add(GetType(context, manager, type));
                var declaration = new Declaration(library.name, file.visibility, DeclarationCategory.Function, compiling.GetChildName(file.name.ToString()), new Tuple(parameterTypes));
                var native = new CompilingNative(file.name, declaration, file.attributes, compiling, cite ? file : null, parameters, new Tuple(returnTypes));
                library.natives.Add(native);
                compiling.AddDeclaration(native);
                file.compiling = native;
            }
        }
    }
}
